File: python/rampidreader/backend.py
"""Адаптер поверх memprocfs/LeechCore. Работает только на Windows с драйвером.

Изолирует ВСЕ вызовы memprocfs, превращая их в чистые объекты, понятные
:mod:`rampidreader.core`. Импорт memprocfs ленивый — модуль грузится только при
реальном открытии устройства, поэтому тесты ядра идут на любой платформе.
"""

from __future__ import annotations

from typing import List, Optional

from .core import Vad


class MemProcFSProcess:
    """Обёртка над объектом процесса memprocfs, реализующая ProcessLike."""

    def __init__(self, vmm_module, proc):
        self._memprocfs = vmm_module
        self._proc = proc
        self.pid = int(proc.pid)
        self.name = str(proc.name)

    def read(self, addr: int, size: int) -> Optional[bytes]:
        try:
            # FLAG_NOCACHE — для «живой» памяти, иначе вернётся закэшированная страница.
            flag = getattr(self._memprocfs, "FLAG_NOCACHE", 0)
            data = self._proc.memory.read(addr, size, flag)
        except Exception:
            return None
        if data is None:
            return None
        return bytes(data)

    def module_base(self, module: str) -> Optional[int]:
        try:
            m = self._proc.module(module)
        except Exception:
            return None
        if m is None:
            return None
        return int(m.base)

    def vads(self) -> List[Vad]:
        try:
            raw = self._proc.maps.vad()
        except Exception:
            return []
        result: List[Vad] = []
        for v in raw:
            # Имена полей в memprocfs могут отличаться между версиями — берём
            # с запасными вариантами и сверим на реальной машине.
            start = _first_attr(v, "va_start", "start", "base", default=0)
            end = _first_attr(v, "va_end", "end", default=0)
            prot = _first_attr(v, "protection", "prot", "flags", default="")
            tag = _first_attr(v, "tag", "info", "name", default="")
            result.append(Vad(int(start), int(end), str(prot), str(tag)))
        return result


def _first_attr(obj, *names, default):
    """Вернуть первый существующий атрибут из ``names`` (поддержка и dict, и объекта)."""
    for n in names:
        if isinstance(obj, dict):
            if n in obj:
                return obj[n]
        elif hasattr(obj, n):
            return getattr(obj, n)
    return default


class MemProcFSBackend:
    """Открывает устройство физпамяти и перечисляет процессы."""

    def __init__(self, device: str, extra_args: Optional[List[str]] = None):
        self.device = device
        self.extra_args = extra_args or []
        self._memprocfs = None
        self._vmm = None

    def open(self) -> "MemProcFSBackend":
        try:
            import memprocfs  # ленивый импорт: только на Windows с драйвером
        except ImportError as e:
            raise RuntimeError(
                "не удалось импортировать memprocfs — установите пакет "
                "(pip install memprocfs) и запустите на Windows с драйвером winpmem"
            ) from e

        self._memprocfs = memprocfs
        args = ["-device", self.device, *self.extra_args]
        try:
            self._vmm = memprocfs.Vmm(args)
        except Exception as e:
            raise RuntimeError(
                f"не удалось инициализировать LeechCore с device={self.device!r}: {e} "
                f"(нужны права администратора и загруженный драйвер)"
            ) from e
        return self

    def processes(self) -> List[MemProcFSProcess]:
        if self._vmm is None:
            raise RuntimeError("backend не открыт — сначала вызовите open()")
        return [
            MemProcFSProcess(self._memprocfs, p) for p in self._vmm.process_list()
        ]

    def close(self) -> None:
        if self._vmm is not None:
            try:
                self._vmm.close()
            except Exception:
                pass
            self._vmm = None

    def __enter__(self) -> "MemProcFSBackend":
        return self.open()

    def __exit__(self, *exc) -> None:
        self.close()

File: rust/src/backend.rs
//! Адаптер поверх крейта `memprocfs` (нативные vmm.dll/leechcore.dll).
//! Компилируется только с фичей `backend`. Изолирует все вызовы memprocfs,
//! превращая их в объекты [`ProcessLike`], понятные чистой логике.

use memprocfs::*;
use rampidreader::{ProcessLike, Vad};

/// Открыть устройство физпамяти. `vmm_dll` — путь к нативной vmm.dll
/// (берём из установленного пакета memprocfs), `device` — строка `PMEM://...`.
pub fn open<'a>(vmm_dll: &str, device: &str) -> Result<Vmm<'a>, String> {
    let args = vec!["-device", device];
    Vmm::new(vmm_dll, &args).map_err(|e| format!("{e}"))
}

/// Обёртка над процессом memprocfs, реализующая [`ProcessLike`].
pub struct MppProcess<'a> {
    proc: VmmProcess<'a>,
    pid: u32,
    name: String,
}

impl<'a> ProcessLike for MppProcess<'a> {
    fn pid(&self) -> u32 {
        self.pid
    }

    fn name(&self) -> &str {
        &self.name
    }

    fn read(&self, addr: u64, size: usize) -> Option<Vec<u8>> {
        // ВАЖНО: mem_read_ex крейта возвращает полный буфер size, ИГНОРИРУЯ число
        // реально прочитанных байт (дыры приходят нулями) — это ломает и пробу, и
        // обнаружение дыр (дамп раздувается до гигабайт). mem_read_into отдаёт
        // фактическое cb_read, как нативный VMMDLL_MemReadEx, — берём его и режем
        // буфер по факту. FLAG_NOCACHE — для «живой» памяти.
        let mut buf = vec![0u8; size];
        match self.proc.mem_read_into(addr, FLAG_NOCACHE, &mut buf) {
            Ok(n) if n > 0 => {
                buf.truncate(n);
                Some(buf)
            }
            _ => None,
        }
    }

    fn module_base(&self, module: &str) -> Option<u64> {
        let mods = self.proc.map_module(false, false).ok()?;
        let target = module.to_lowercase();
        mods.into_iter()
            .find(|m| m.name.to_lowercase() == target)
            .map(|m| m.va_base)
    }

    fn vads(&self) -> Vec<Vad> {
        match self.proc.map_vad(true) {
            Ok(list) => list
                .into_iter()
                .map(|v| {
                    let prot = if v.is_mem_commit { "commit" } else { "reserve" };
                    Vad::new(v.va_start, v.va_end, prot, &v.info)
                })
                .collect(),
            Err(_) => Vec::new(),
        }
    }
}

/// Перечислить процессы устройства.
pub fn processes<'a>(vmm: &'a Vmm<'a>) -> Vec<MppProcess<'a>> {
    let mut out = Vec::new();
    if let Ok(list) = vmm.process_list() {
        for p in list {
            let pid = p.pid;
            let name = p.info().map(|i| i.name).unwrap_or_default();
            out.push(MppProcess { proc: p, pid, name });
        }
    }
    out
}
